src/chunk_adaptive.c

```diff
@@ -0,0 +1,551 @@
+			found = minmax_indexscan(rel, idxrel, attnum, minmax);
+
+		index_close(idxrel, AccessShareLock);
+
+		if (found)
+			break;
+	}
+
+	if (!found)
+		found = minmax_heapscan(rel, atttype, attnum, minmax);
+
+	heap_close(rel, AccessShareLock);
+
+	return found;
+}
+
+#define CHUNK_SIZING_FUNC_NARGS 3
+#define DEFAULT_CHUNK_WINDOW 3
+
+/* Tuples must have filled this fraction of the chunk interval to use it to
+ * estimate a new chunk time interval */
+#define INTERVAL_FILLFACTOR_THRESH 0.5
+/* A chunk must fill this (extrapolated) fraction of the target size to use it
+ * to estimate a new chunk time interval.  */
+#define SIZE_FILLFACTOR_THRESH 0.15
+
+/* The calculated chunk time interval must differ this much to actually change
+ * the interval */
+#define INTERVAL_MIN_CHANGE_THRESH 0.01
+
+TS_FUNCTION_INFO_V1(calculate_chunk_interval);
+
+Datum
+calculate_chunk_interval(PG_FUNCTION_ARGS)
+{
+	int32		dimension_id = PG_GETARG_INT32(0);
+	int64		dimension_coord = PG_GETARG_INT64(1);
+	int64		chunk_target_size_bytes = PG_GETARG_INT64(2);
+	int64		chunk_interval = 0;
+	int32		hypertable_id;
+	Hypertable *ht;
+	Dimension  *dim;
+	List	   *chunks = NIL;
+	ListCell   *lc;
+	int			num_intervals = 0;
+	double		interval_diff;
+
+	if (PG_NARGS() != CHUNK_SIZING_FUNC_NARGS)
+		elog(ERROR, "invalid number of arguments");
+
+	hypertable_id = dimension_get_hypertable_id(dimension_id);
+
+	if (hypertable_id <= 0)
+		elog(ERROR, "could not find a matching hypertable for dimension %u", dimension_id);
+
+	ht = hypertable_get_by_id(hypertable_id);
+
+	Assert(ht != NULL);
+
+	dim = hyperspace_get_dimension_by_id(ht->space, dimension_id);
+
+	Assert(dim != NULL);
+
+	/* Get a window of recent chunks */
+	chunks = chunk_get_window(hypertable_id, dimension_coord, DEFAULT_CHUNK_WINDOW);
+
+	foreach(lc, chunks)
+	{
+		Chunk	   *chunk = lfirst(lc);
+		DimensionSlice *slice = hypercube_get_slice_by_dimension_id(chunk->cube, dimension_id);
+		int64		chunk_size,
+					slice_interval;
+		Datum		minmax[2];
+
+		Assert(NULL != slice);
+
+		chunk_size = DatumGetInt64(DirectFunctionCall1(pg_total_relation_size,
+													   ObjectIdGetDatum(chunk->table_id)));
+
+		slice_interval = slice->fd.range_end - slice->fd.range_start;
+
+		if (chunk_get_minmax(chunk->table_id, dim->fd.column_type, dim->column_attno, minmax))
+		{
+			int64		min = time_value_to_internal(minmax[0], dim->fd.column_type);
+			int64		max = time_value_to_internal(minmax[1], dim->fd.column_type);
+			double		interval_fillfactor,
+						size_fillfactor;
+			int64		extrapolated_chunk_size;
+
+			/*
+			 * The fillfactor of the slice interval that the data actually
+			 * spans
+			 */
+			interval_fillfactor = ((double) max - min) / slice_interval;
+
+			/*
+			 * Extrapolate the size the chunk would have if it spanned the
+			 * entire interval
+			 */
+			extrapolated_chunk_size = chunk_size / interval_fillfactor;
+			size_fillfactor = ((double) extrapolated_chunk_size) / chunk_target_size_bytes;
+
```

Comment:
Can we have all these values be print out into the log at DEBUG1 or something. Right now this is totally a black box.

Reply:
Maybe even better to have a guc to define whether to output or not.